lib/Runtime/PlatformAgnostic/ICU.h

```diff
@@ -22,65 +22,47 @@ namespace PlatformAgnostic
 {
     namespace ICUHelpers
     {
-        template<typename T>
-        class AutoICUObject
+        template<typename TObject, void(__cdecl * CloseFunction)(TObject)>
+        class ScopedICUObject
         {
-        public:
-            typedef void(__cdecl * CloseFunction)(T);
         private:
-            T object;
-            CloseFunction closeFunc;
+            TObject object;
         public:
-            explicit AutoICUObject(T object);
-            AutoICUObject(const AutoICUObject &other) = delete;
-            ~AutoICUObject();
-            operator T();
-            T Detatch();
-        };
-
-        template<typename T>
-        constexpr typename AutoICUObject<T>::CloseFunction CloseFunctionForType(T type);
-
-        template<typename T>
-        AutoICUObject<T>::AutoICUObject(T object) : object(object)
-        {
-            closeFunc = CloseFunctionForType(object);
-        }
-
-        template<typename T>
-        AutoICUObject<T>::~AutoICUObject()
-        {
-            if (object != nullptr)
+            // The constructor is implemented later by INSTANTIATE_SCOPEDICUOBJECT
+            explicit ScopedICUObject(TObject object);
+            ScopedICUObject(const ScopedICUObject &other) = delete;
+            ScopedICUObject(ScopedICUObject &&other) = delete;
+            ScopedICUObject &operator=(const ScopedICUObject &other) = delete;
+            ScopedICUObject &operator=(ScopedICUObject &&other) = delete;
+            ~ScopedICUObject()
             {
-                closeFunc(object);
+                if (object != nullptr)
+                {
+                    CloseFunction(object);
+                }
             }
-        }
-
-        template<typename T>
-        AutoICUObject<T>::operator T()
-        {
-            return object;
-        }
 
-        template<typename T>
-        T AutoICUObject<T>::Detatch()
-        {
-            T ret = object;
-            object = nullptr;
-            return ret;
-        }
+            operator TObject()
+            {
+                return object;
+            }
+        };
+    }
+}
 
-#define INSTANTIATE_AUTOICUOBJECT(T, Closer) \
+#define INSTANTIATE_SCOPEDICUOBJECT(TObject, CloseFunction) \
         template<> \
-        constexpr AutoICUObject<T *>::CloseFunction CloseFunctionForType(T *) \
+        inline PlatformAgnostic::ICUHelpers::ScopedICUObject<TObject *, CloseFunction>::ScopedICUObject(TObject * object) : \
+            object(object) \
```

Comment:
Any particular reason these constructors need to be defined in macros rather than in the main class template?

Reply:
Just realized that they don't anymore -- the template parameter is enough, which makes this all actually much cleaner. Thanks!